enterprise/server/auditlog/auditlog.go

```diff
@@ -0,0 +1,263 @@
+	// Map of FooState protos to their corresponding fields in ResourceState proto.
+	payloadTypes map[protoreflect.MessageDescriptor]protoreflect.FieldDescriptor
+}
+
+func GroupAPIKeyResourceID(id, name string) *alpb.ResourceID {
+	return &alpb.ResourceID{
+		Type: alpb.ResourceType_GROUP_API_KEY,
+		Id:   id,
+		Name: name,
+	}
+}
+
+func UserAPIKeyResourceID(id, name string) *alpb.ResourceID {
+	return &alpb.ResourceID{
+		Type: alpb.ResourceType_USER_API_KEY,
+		Id:   id,
+		Name: name,
+	}
+}
+
+func GroupResourceID(id string) *alpb.ResourceID {
+	return &alpb.ResourceID{
+		Type: alpb.ResourceType_GROUP,
+		Id:   id,
+	}
+}
+
+func SecretResourceID(secretName string) *alpb.ResourceID {
+	return &alpb.ResourceID{
+		Type: alpb.ResourceType_SECRET,
+		Id:   secretName,
+	}
+}
+
+func InvocationResourceID(id string) *alpb.ResourceID {
+	return &alpb.ResourceID{
+		Type: alpb.ResourceType_INVOCATION,
+		Id:   id,
+	}
+}
+
+func Register(env environment.Env) error {
+	if !*auditLogsEnabled {
+		return nil
+	}
+
+	if env.GetOLAPDBHandle() == nil {
+		return status.FailedPreconditionErrorf("audit logs require an OLAP database")
+	}
+
+	payloadTypes := make(map[protoreflect.MessageDescriptor]protoreflect.FieldDescriptor)
+	pfs := (&alpb.Entry_ResourceRequest{}).ProtoReflect().Descriptor().Fields()
+	for i := 0; i < pfs.Len(); i++ {
+		pf := pfs.Get(i)
+		payloadTypes[pf.Message()] = pf
+	}
+	l := &Logger{
+		env:          env,
+		dbh:          env.GetOLAPDBHandle(),
+		payloadTypes: payloadTypes,
+	}
+	env.SetAuditLog(l)
+	return nil
+}
+
+// wrapRequestProto automatically finds and sets the correct child message of
+// the ResourceState proto based on the type of the passed proto.
+func (l *Logger) wrapRequestProto(payload proto.Message) (*alpb.Entry_ResourceRequest, error) {
+	fd, ok := l.payloadTypes[payload.ProtoReflect().Descriptor()]
+	if !ok {
+		return nil, status.InvalidArgumentErrorf("invalid payload proto: %s", payload.ProtoReflect().Descriptor())
+	}
+	payloadWrapper := &alpb.Entry_ResourceRequest{}
+	payloadWrapper.ProtoReflect().Set(fd, protoreflect.ValueOfMessage(payload.ProtoReflect()))
+	return payloadWrapper, nil
+}
+
+func clearRequestContext(request proto.Message) proto.Message {
+	fd := request.ProtoReflect().Descriptor().Fields().ByName("request_context")
+	if fd == nil {
+		return request
+	}
+	request = proto.Clone(request)
+	request.ProtoReflect().Clear(fd)
+	return request
+}
+
+func (l *Logger) insertLog(ctx context.Context, resource *alpb.ResourceID, method string, request proto.Message) error {
+	u, err := l.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return status.WrapError(err, "auth failed")
+	}
+	if u.GetUserID() == "" {
+		return nil
+	}
+
+	ui, err := l.env.GetUserDB().GetUser(ctx)
+	if err != nil {
+		return status.WrapError(err, "could not lookup user")
+	}
+
+	request = clearRequestContext(request)
+
+	var requestBytes []byte
+	if request != nil {
+		rp, err := l.wrapRequestProto(request)
```

Comment:
idk if you want to do any validation here to ensure that the request proto type matches the resource ID type (and bail if it does not)